### DISCUSSION TITLE
Custom validation of Query parameters using Pydantic models

### DISCUSSION THREAD

**Discussion by dfredouille-dg** [fastapi/fastapi#9518] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import datetime
from fastapi import Depends, FastAPI, HTTPException, status
from pydantic import BaseModel, ValidationError, validator

app = FastAPI()

class TimeRange(BaseModel):
    start_at: datetime 
    end_at: datetime

    @validator("end_at")
    def end_must_be_strictly_after_start(cls, v, values):
        if "start_at" in values and v <= values["start_at"]:
            raise ValueError("End timestamp must be strictly after start timestamp")
        return v

def with_timerange(start_at: datetime, end_at: datetime):
    try:
        return TimeRange(start_at=start_at, end_at=end_at)

    except ValidationError as exc:
        error_details = [
            dict(
                loc=["query"] + [loc for loc in error["loc"]],
                msg=error["msg"],
                type=f"timerange.{error['type']}",
            )
            for error in exc.errors()
        ]
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY, detail=error_details
        )

@app.get("/")
async def root422(
    tr: TimeRange = Depends(with_timerange),
):
    return {"tr": tr}
```

### Description

This is about help improving my code, and perhaps asking for a related improvement in `fastapi`. 

* My aim: In the example above, the TimeRange class verifies that the end_at is after the start_at and raises a ValidationError if it is not the case. I can (and do - code not included) use this in a body and if there is a client input not respecting it I get a nice 422 with error details containing location, type and message. I want to validate query parameter inputs as easily as I can do it for body content, i.e., using Pydandic, and obtaining a nice 422 response. This is what the `with_timerange()` logic does in the above example.
* The problem: I feel like I am reinventing the wheel. The code that does the conversion from ValidationError to HttpException(422) likely already exists in `fastapi` somehow (but I did not find where and/or if I could reuse it). 

Potential solutions I can / would love to see:
* Having a method from `fastapi` that does the conversion logic for me (I understand that you may not want to return a 422 on all ValidationErrors as some may not be related to the input), but providing a mechanism for doing so would be very nice.
* Have a way to inject the Pydantic model validator by declaring it directly in the endpoint definition, a very crude first idea would be something like.
```
@app.get("/")
async def root422(
    tr: TimeRange = Depends(with_timerange, is_parameter_validator=True),
):
```

Side note: Searching for related discussions, I found https://github.com/tiangolo/fastapi/discussions/7919 and https://github.com/tiangolo/fastapi/discussions/8143. They want to build models automatically which would be nice too. however on my side the case is simpler: I ask to still build the model by hand with a dependency but get an easier way to handle errors.

### Operating System

macOS

### Operating System Details

MacOS 13.2.1 (22D68) with intel core

### FastAPI Version

0.95.1

### Python Version

3.11.1

**Comment by YuriiMotov** [comment, marked as answer]
Since V 0.115.0 FastAPI officially supports Pydantic models to define `Query` parameters!

Here is how it can be implemented now:
```python
from datetime import datetime
from typing import Any
from fastapi import FastAPI, Query
from pydantic import BaseModel, ValidationInfo, field_validator

app = FastAPI()


class TimeRange(BaseModel):
    start_at: datetime
    end_at: datetime

    @field_validator("end_at")
    @classmethod
    def end_must_be_strictly_after_start(cls, v: Any, info: ValidationInfo):
        data = info.data
        if "start_at" in data and v <= data["start_at"]:
            raise ValueError("End timestamp must be strictly after start timestamp")
        return v


@app.get("/")
async def root422(tr: TimeRange = Query()):
    return {"tr": tr}
```
